src/Microsoft.ML.Transforms/CustomMappingTransformerFactory.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.Data.DataView;
+
+namespace Microsoft.ML.Transforms
+{
+    /// <summary>
+    /// Place this attribute onto a type to cause it to be considered a custom mapping transformer factory.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class)]
+    public sealed class CustomMappingTransformerFactoryAttribute : ExtensionBaseAttribute
+    {
+        public CustomMappingTransformerFactoryAttribute(string contractName)
+            : base(contractName)
+        {
+        }
+    }
+
+    internal interface ICustomMappingTransformerFactory
+    {
+        ITransformer CreateTransformerObject(IHostEnvironment env, string contractName);
+    }
+
+    /// <summary>
+    /// The base type for custom mapping transformer factories.
+    /// </summary>
+    /// <typeparam name="TSrc">The type that describes what 'source' columns are consumed from the input <see cref="IDataView"/>.</typeparam>
+    /// <typeparam name="TDst">The type that describes what new columns are added by this transform.</typeparam>
+    public abstract class CustomMappingTransformerFactory<TSrc, TDst> : ICustomMappingTransformerFactory
```

Comment:
One of the things that I'm not 100% satisfied with yet is the names I used here. I wonder if we should drop the `er` and just be `CustomMappingTransformFactory` and `public abstract Action<TSrc, TDst> GetTransform();`.

One reason for this is that the things this creates aren't actually `Transformer`s, but instead just actions that the actual `Transformer` will call.

Reply:
We've elsewhere called the things that aren't transformers but the simpler things that transformers *call* to do the transformation mappers... e.g., `IRowToRowMapper`, `ISchemaBindableMapper`, `IValueMapper`... now granted many of these are not public concepts and some of them don't quite map onto what you're doing, but if we think about this, we might consider that the thing you're creating in this factory is not the `CustomMappingTransformer`, but just the mapping. So maybe `ICustomMappingFactory` or `ICustomMapperFactory` might be the better choice, as it is more consistent with what we do elsewhere? Not insisting on that particular solution, just wondering if that helps clear your logjam to know how I would rename it.

However, just to be clear, while I don't insist on the above solution, I do feel strongly there needs to be a different name to what it is now. Calling this a "transformer factory" with the abstract method `GetTransformer` is inappropriate... `Transformer` is already a very well established central concept in this library, and this method definitely is *not* a factory for them, and must not advertise itself as such.

---
In reply to: [257331335](https://github.com/dotnet/machinelearning/pull/2569#discussion_r257331335) [](ancestors = 257331335)